folding-schemes-solidity/templates/kzg10_verifier.askama.sol

```diff
@@ -0,0 +1,263 @@
+
+    /**
+     * @notice  Performs scalar multiplication in G_1.
+     * @param   p  G_1 point to multiply
+     * @param   s  Scalar to multiply by
+     * @return  r  G_1 point p multiplied by scalar s
+     */
+    function mulScalar(uint256[2] memory p, uint256 s) internal view returns (uint256[2] memory r) {
+        uint256[3] memory input;
+        input[0] = p[0];
+        input[1] = p[1];
+        input[2] = s;
+        bool success;
+        assembly {
+            success := staticcall(sub(gas(), 2000), 7, input, 0x80, r, 0x60)
+            switch success
+            case 0 { invalid() }
+        }
+        require(success, "bn254: scalar mul failed");
+    }
+
+    /**
+     * @notice  Negates a point in G_1.
+     * @param   p  G_1 point to negate
+     * @return  uint256[2]  G_1 point -p
+     */
+    function negate(uint256[2] memory p) internal pure returns (uint256[2] memory) {
+        if (p[0] == 0 && p[1] == 0) {
+            return p;
+        }
+        return [p[0], BN254_PRIME_FIELD - (p[1] % BN254_PRIME_FIELD)];
+    }
+
+    /**
+     * @notice  Adds two points in G_1.
+     * @param   p1  G_1 point 1
+     * @param   p2  G_1 point 2
+     * @return  r  G_1 point p1 + p2
+     */
+    function add(uint256[2] memory p1, uint256[2] memory p2) internal view returns (uint256[2] memory r) {
+        bool success;
+        uint256[4] memory input = [p1[0], p1[1], p2[0], p2[1]];
+        assembly {
+            success := staticcall(sub(gas(), 2000), 6, input, 0xc0, r, 0x60)
+            switch success
+            case 0 { invalid() }
+        }
+
+        require(success, "bn254: point add failed");
+    }
+
+    /**
+     * @notice  Computes the pairing check e(p1, p2) * e(p3, p4) == 1
+     * @dev     Note that G_2 points a*i + b are encoded as two elements of F_p, (a, b)
+     * @param   a_1  G_1 point 1
+     * @param   a_2  G_2 point 1
+     * @param   b_1  G_1 point 2
+     * @param   b_2  G_2 point 2
+     * @return  result  true if pairing check is successful
+     */
+    function pairing(uint256[2] memory a_1, uint256[2][2] memory a_2, uint256[2] memory b_1, uint256[2][2] memory b_2)
+        internal
+        view
+        returns (bool result)
+    {
+        uint256[12] memory input = [
+            a_1[0],
+            a_1[1],
+            a_2[0][1], // imaginary part first
+            a_2[0][0],
+            a_2[1][1], // imaginary part first
+            a_2[1][0],
+            b_1[0],
+            b_1[1],
+            b_2[0][1], // imaginary part first
+            b_2[0][0],
+            b_2[1][1], // imaginary part first
+            b_2[1][0]
+        ];
+
+        uint256[1] memory out;
+        bool success;
+
+        assembly {
+            success := staticcall(sub(gas(), 2000), 8, input, 0x180, out, 0x20)
+            switch success
+            case 0 { invalid() }
+        }
+
+        require(success, "bn254: pairing failed");
+
+        return out[0] != 0;
```

Comment:
I feel `return out[0] == 1;` will be more straightforward, but it's fine as is.

Reply:
good suggestion! [478d9f2](https://github.com/privacy-scaling-explorations/folding-schemes/pull/70/commits/478d9f280ee1b8b5c201fed8730f03acf6ecbb6a)!